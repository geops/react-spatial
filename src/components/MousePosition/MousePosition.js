import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import OLMap from 'ol/Map';
import { createStringXY } from 'ol/coordinate';
import OLMousePosition from 'ol/control/MousePosition';
import Select from '../Select';

const propTypes = {
  /**
   * The current map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * CSS class of the checkbox.
   */
  className: PropTypes.string,

  /**
   * List of projections to display.
   */
  projections: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The label to display in the select box.
       */
      label: PropTypes.string.isRequired,

      /**
       * The value used to create the optionsÂ´s projection of the MousePosition control.
       * See [doc](https://openlayers.org/en/latest/apidoc/module-ol_control_MousePosition.html).
       */
      value: PropTypes.string.isRequired,

      /**
       * A function following the  [CoordinateFormat](https://openlayers.org/en/latest/apidoc/module-ol_coordinate.html#~CoordinateFormat).
       */
      format: PropTypes.func,
    }),
  ),
};

const defaultProps = {
  className: 'tm-mouse-position',
  projections: [
    {
      label: 'EPSG:4326',
      value: 'EPSG:4326',
    },
    {
      label: 'EPSG:3857',
      value: 'EPSG:3857',
    },
  ],
};

class MousePosition extends PureComponent {
  constructor(props) {
    super(props);
    const { projections } = this.props;
    this.state = {
      projection: projections && projections[0],
    };
    this.ref = React.createRef();
  }

  componentDidMount() {
    this.updateControl();
  }

  componentDidUpdate(prevProps, prevState) {
    const { projection } = this.state;

    if (prevState.projection !== projection) {
      this.updateControl();
    }
  }

  componentWillUnmount() {
    if (this.control) {
      const { map } = this.props;
      map.removeControl(this.control);
    }
  }

  updateControl() {
    const { map } = this.props;
    const { projection } = this.state;

    if (this.control) {
      map.removeControl(this.control);
    }

    if (!projection || !this.ref || !this.ref.current) {
      return;
    }

    this.control = new OLMousePosition({
      coordinateFormat: projection.format || createStringXY(4),
      projection: projection.value,
      target: this.ref.current,
      undefinedHTML: '&nbsp;',
    });
    map.addControl(this.control);
  }

  renderSelect() {
    const { projections } = this.props;
    if (!projections.length) {
      return null;
    }

    return (
      <Select
        options={projections}
        onChange={(evt, val) => {
          this.setState({
            projection: projections.find(p => p.value === val),
          });
        }}
      />
    );
  }

  render() {
    const { className } = this.props;
    return (
      <div className={className}>
        {this.renderSelect()}
        <span ref={this.ref} />
      </div>
    );
  }
}

MousePosition.propTypes = propTypes;
MousePosition.defaultProps = defaultProps;

export default MousePosition;
