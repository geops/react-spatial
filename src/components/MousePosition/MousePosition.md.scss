.rs-mouse-position-example {
  .rs-mouse-position {
    display: flex;
    align-items: center;
    margin: 10px 0;

    span {
      margin-left: 10px;
    }

    select {
      background: none;
      height: 20px;
    }
  }
}
