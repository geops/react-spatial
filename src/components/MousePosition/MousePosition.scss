.tm-mouse-position {
  & > select {
    display: inline-block;
  }
}
