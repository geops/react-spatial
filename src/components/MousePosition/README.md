#

This demonstrates the use of MousePosition.

```jsx
import React from  'react';
import OLMap from 'ol/Map';
import BasicMap from 'react-spatial/components/BasicMap';
import LayerService from 'react-spatial/LayerService';
import ConfigReader from 'react-spatial/ConfigReader';
import MousePosition from 'react-spatial/components/MousePosition';

class MousePositionExample extends React.Component {
  constructor(props) {
    super(props);
     this.center = [1149722.7037660484, 6618091.313553318];
    this.map = new OLMap({controls:[]});
    this.state = {
      layerService: null
    }
  }

  componentDidMount() {
    const layers = ConfigReader.readConfig(
      this.map,
      treeData,
    );
    this.setState({
      layerService: new LayerService(layers)
    })
  }

  render() {
    return (
      <div className="tm-mouse-position-example">
        <BasicMap map={this.map} center={this.center} zoom={6} />
        <MousePosition map={this.map} projections={[
        {
          label: 'World Geodetic System 1984',
          value: 'EPSG:4326',
        },
        {
          label: 'Web mercator with custom output',
          value: 'EPSG:3857',
          format: (coord) => {
            const decimals = 4;
            let text = [];

            coord.forEach((input)=> {
                input = Math.round(parseFloat(input) * Math.pow(10, decimals)) /
                    Math.pow(10, decimals);
                input =input.toString();
                var parts = input.toString().split('.');
                parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, "'");
                text.push(parts.join());
            });
            return 'Coordinates: ' + text.join(' ');
          }
        },]}/>
      </div>
    )
  }
}

<MousePositionExample />;
```
