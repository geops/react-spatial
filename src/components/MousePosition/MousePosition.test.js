import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-canvas-mock';
import OLMap from 'ol/Map';
import OLMousePosition from 'ol/control/MousePosition';
import MousePosition from './MousePosition';

configure({ adapter: new Adapter() });
const expectSnapshot = props => {
  const map = new OLMap({});
  const component = renderer.create(<MousePosition map={map} {...props} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
};

describe('MousePosition', () => {
  describe('matches snapshots', () => {
    test('using default values.', () => {
      expectSnapshot({});
    });

    test('using no projections.', () => {
      expectSnapshot({
        projections: [],
      });
    });

    test('using only one projection', () => {
      expectSnapshot({
        projections: [{ label: 'foo', value: 'foo' }],
      });
    });

    test('using multiple projections.', () => {
      expectSnapshot({
        projections: [
          { label: 'foo', value: 'foo' },
          { label: 'bar', value: 'bar' },
        ],
      });
    });
  });

  test('add MousePosition control to the map.', () => {
    const map = new OLMap({});
    const spy = jest.spyOn(map, 'removeControl');
    const spy2 = jest.spyOn(map, 'addControl');
    const fn = jest.fn();
    const wrapper = mount(
      <MousePosition
        map={map}
        projections={[
          {
            label: 'EPSG:4326',
            value: 'EPSG:4326',
            format: fn,
          },
        ]}
      />,
    );
    const ctrl = wrapper.instance().control;
    expect(spy).toHaveBeenCalledTimes(0);
    expect(spy2).toHaveBeenCalledTimes(1);
    expect(spy2.mock.calls[0][0]).toBeInstanceOf(OLMousePosition);
    expect(spy2.mock.calls[0][0]).toBe(ctrl);
    expect(ctrl.getProjection().getCode()).toBe('EPSG:4326');
    expect(ctrl.getCoordinateFormat()).toBe(fn);
  });

  test('remove MousePosition control before adding it again.', () => {
    const map = new OLMap({});
    const spy = jest.spyOn(map, 'removeControl');
    const spy2 = jest.spyOn(map, 'addControl');
    const wrapper = mount(<MousePosition map={map} />);
    expect(spy).toHaveBeenCalledTimes(0);
    expect(spy2).toHaveBeenCalledTimes(1);
    wrapper.setState({
      projection: wrapper.props().projections[1],
    });
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0]).toBe(spy2.mock.calls[0][0]);
    expect(spy2).toHaveBeenCalledTimes(2);
    expect(spy2.mock.calls[1][0]).toBe(wrapper.instance().control);
    expect(spy2.mock.calls[1][0]).toBeInstanceOf(OLMousePosition);
  });

  test('remove MousePosition control on unmount.', () => {
    const map = new OLMap({});
    const spy = jest.spyOn(map, 'removeControl');
    const spy2 = jest.spyOn(map, 'addControl');
    const wrapper = mount(<MousePosition map={map} />);
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.unmount();
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0]).toBe(spy2.mock.calls[0][0]);
  });
});
