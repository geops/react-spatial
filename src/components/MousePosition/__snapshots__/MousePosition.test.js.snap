// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MousePosition matches snapshots using default values. 1`] = `
<div
  className="tm-mouse-position"
>
  <select
    className="tm-select"
    onChange={[Function]}
  >
    <option
      value="EPSG:4326"
    >
      EPSG:4326
    </option>
    <option
      value="EPSG:3857"
    >
      EPSG:3857
    </option>
  </select>
  <span />
</div>
`;

exports[`MousePosition matches snapshots using multiple projections. 1`] = `
<div
  className="tm-mouse-position"
>
  <select
    className="tm-select"
    onChange={[Function]}
  >
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
  </select>
  <span />
</div>
`;

exports[`MousePosition matches snapshots using no projections. 1`] = `
<div
  className="tm-mouse-position"
>
  <span />
</div>
`;

exports[`MousePosition matches snapshots using only one projection 1`] = `
<div
  className="tm-mouse-position"
>
  <select
    className="tm-select"
    onChange={[Function]}
  >
    <option
      value="foo"
    >
      foo
    </option>
  </select>
  <span />
</div>
`;
